packages/components/src/disabled/index.tsx

```diff
@@ -0,0 +1,82 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import type { RefCallback } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { useDisabled } from '@wordpress/compose';
+import { createContext } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { StyledWrapper } from './styles/disabled-styles';
+import type { DisabledProps } from './types';
+import type { WordPressComponentProps } from '../ui/context';
+
+const Context = createContext< boolean >( false );
+const { Consumer, Provider } = Context;
+
+/**
+ * `Disabled` is a component which disables descendant tabbable elements and prevents pointer interaction.
+ *
+ * @example
+ * ```js
+ * import { Button, Disabled, TextControl } from '@wordpress/components';
+ * import { useState } from '@wordpress/element';
+ *
+ * const MyDisabled = () => {
+ * 	const [ isDisabled, setIsDisabled ] = useState( true );
+ *
+ * 	let input = <TextControl label="Input" onChange={ () => {} } />;
+ * 	if ( isDisabled ) {
+ * 		input = <Disabled>{ input }</Disabled>;
+ * 	}
+ *
+ * 	const toggleDisabled = () => {
+ * 		setIsDisabled( ( state ) => ! state );
+ * 	};
+ *
+ * 	return (
+ * 		<div>
+ * 			{ input }
+ * 			<Button variant="primary" onClick={ toggleDisabled }>
+ * 				Toggle Disabled
+ * 			</Button>
+ * 		</div>
+ * 	);
+ * };
+ * ```
+ */
+function Disabled( {
+	className,
+	children,
+	isDisabled = true,
+	...props
+}: WordPressComponentProps< DisabledProps, 'div' > ) {
+	const ref: RefCallback< HTMLElement > = useDisabled();
```

Comment:
Since `useDisabled` (and the underlying `useRefEffect` hook) are typed, there shouldn't be a need to explicitly add a type to `ref` here:

```suggestion
	const ref = useDisabled();
```

unless we want to be specific about the `ref` targeting a `div` element (instead of a generic HTML element), similarly to how `ref` was typed before this PR:

```suggestion
	const ref: RefCallback< HTMLDivElement > = useDisabled();
```